### PR TITLE
Make comments work

### LINKED ISSUE #4170: Make comments work.
We should check:
 * Which states comments are allowed in;
 * Whether the person is allowed to comment;
 * That comments are displayed according to their visibility;

### PR DESCRIPTION
**Pull request by mic4ael**
closes #4170

### CODE REVIEW

File: indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx [hunk 1]
```diff
@@ -58,20 +63,28 @@ export default function ReviewForm({block}) {
 
   const InputComponent = commentFormVisible ? FinalTextArea : FinalInput;
   const inputProps = commentFormVisible ? {autoFocus: true} : {};
+  const addComment = async formData => {
+    if (commentFormVisible) {
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx]: what's the reason for this check? can you even submit the form when it's not visible?

File: indico/modules/events/editing/models/editable.py [hunk 2]
```diff
@@ -77,3 +79,16 @@ def __repr__(self):
...
     # TODO: state - either a column property referencing the newest revision's state or a normal column
+
+    def can_comment(self, user):
+        return self.editor == user or self.contribution.is_user_associated(user, check_abstract=True)
+
+    @property
+    def editors(self):
+        editors = [self.editor] if self.editor else []
+        if session.user not in editors and self.contribution.is_user_associated(session.user, check_abstract=True):
+            editors.append(session.user)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/models/editable.py]: this logic feels super weird. also, accessing the session from a model?

File: indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx (hunk 1 above)
Comment (mic4ael) [reply indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx]: Nope, it shouldn't be possible. I just put it here for precaution (even thouigh it should never happen)

File: indico/modules/events/editing/models/editable.py (hunk 2 above)
Comment (mic4ael) [reply indico/modules/events/editing/models/editable.py]: As to this one, let's discuss it in person.

File: indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx [hunk 3]
```diff
@@ -43,6 +46,10 @@ const judgmentOptions = [
 ];
 
 export default function ReviewForm({block}) {
+  const dispatch = useDispatch();
+  const {contributionId, eventId, editableType} = useSelector(getStaticData);
+  const {editors} = useSelector(getDetails);
+  const isEditor = editors.find(editor => editor.id === Indico.User.id) !== undefined;
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx]: when would you have more than one editor on a revision?

also, i would send explicit permissions back from the server

File: indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx [hunk 4]
```diff
@@ -58,31 +65,48 @@ export default function ReviewForm({block}) {
 
   const InputComponent = commentFormVisible ? FinalTextArea : FinalInput;
   const inputProps = commentFormVisible ? {autoFocus: true} : {};
+  const addComment = async (formData, form) => {
+    const rv = await dispatch(
+      createRevisionComment(eventId, contributionId, editableType, block.id, formData)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx]: considering how much data we are passing here just to build the URL I wonder if adding a `create_comment_url` in the marshmallow schema wouldn't be easier

File: indico/modules/events/editing/client/js/components/Timeline/index.jsx [hunk 5]
```diff
@@ -19,7 +19,7 @@ import TimelineItem from './TimelineItem';
 export default function Timeline() {
   const dispatch = useDispatch();
   const details = useSelector(selectors.getDetails);
-  const isLoading = useSelector(selectors.isLoading);
+  const isInitialPaperDetailsLoading = useSelector(selectors.isInitialPaperDetailsLoading);
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/index.jsx]: s/Paper/Editable/

File: indico/modules/events/editing/models/editable.py [hunk 6]
```diff
@@ -77,3 +77,13 @@ def __repr__(self):
...
 
     # TODO: state - either a column property referencing the newest revision's state or a normal column
+
+    def can_comment(self, user):
+        return self.editor == user or self.contribution.is_user_associated(user, check_abstract=True)
+
+    @property
+    def editors(self):
+        editors = [self.editor] if self.editor else []
+        editors.extend([revision.editor for revision in self.revisions
```
Comment (ThiefMaster) [inline indico/modules/events/editing/models/editable.py]: you don't need the `[]` here

File: indico/modules/events/editing/schemas.py [hunk 7]
```diff
@@ -79,19 +79,35 @@ class Meta:
...
     final_state = fields.Nested(RevisionState)
 
+    def _get_comments(self, revision):
+        comments = []
+        editors = self.context.get('editors', [])
+        current_user = self.context.get('user')
+
+        for comment in revision.comments:
+            if not comment.internal or current_user in editors:
+                comments.append(comment)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/schemas.py]: you can use a list comprehension here

File: indico/modules/events/editing/client/js/components/Timeline/CommentItem.jsx [hunk 8]
```diff
@@ -6,24 +6,74 @@
...
-export default function Comment({user, createdDt, modifiedDt, html}) {
+import {deleteRevisionComment} from '../../actions';
+import {getDetails, getStaticData} from '../../selectors';
+
+const INDICO_BOT_USER = {
+  fullName: 'Indico Bot',
+  avatarBgColor: '#8f8f8f',
+};
+
+function reducer(state, action) {
```
[the remark this answers is not in the document]
Comment (mic4ael) [reply indico/modules/events/editing/client/js/components/Timeline/CommentItem.jsx]: Yeah, not sure why I didn't use `useState` first :P (Maybe too many times I handled request-related stuff with redux)

File: indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx (hunk 4 above)
Comment (mic4ael) [reply indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx]: Yeah, maybe that is a good idea so I wouldn't have to pass additional props between components

File: indico/modules/events/editing/models/comments.py [hunk 9]
```diff
@@ -100,3 +101,24 @@ class EditingRevisionComment(RenderModeMixin, db.Model):
     @return_ascii
     def __repr__(self):
         return format_repr(self, 'id', 'revision_id', 'user_id', internal=False, _text=text_to_repr(self.text))
+
+    @locator_property
+    def locator(self):
+        return {
```
Comment (mic4ael) [inline indico/modules/events/editing/models/comments.py]: This is only temporary until one of my PRs is merged

File: indico/modules/events/editing/models/editable.py [hunk 10]
```diff
@@ -82,3 +82,13 @@ def __repr__(self):
     @locator_property
     def locator(self):
         return dict(self.contribution.locator, type=self.type.name)
+
+    def can_comment(self, user):
+        return self.editor == user or self.contribution.is_user_associated(user, check_abstract=True)
+
+    @property
+    def editors(self):
```
Comment (ThiefMaster) [inline indico/modules/events/editing/models/editable.py]: I think we want to check only the `editor` assigned to the Editable, at least for anything that modifies data.

Not sure about creating new internal comments, I guess it'd make sense to let all editors (based on the not-yet-existing event permission, and the historical list of editors on the revision) create internal comments.

Comment (mic4ael) [reply indico/modules/events/editing/models/editable.py]: @meluru opinions?

File: indico/modules/events/editing/models/editable.py [hunk 11]
```diff
@@ -82,3 +82,6 @@ def __repr__(self):
     @locator_property
     def locator(self):
         return dict(self.contribution.locator, type=self.type.name)
+
+    def can_comment(self, user):
+        return self.editor == user or self.contribution.is_user_associated(user, check_abstract=True)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/models/editable.py]: the first check should use the new event permission instead

File: indico/modules/events/editing/schemas.py [hunk 12]
```diff
@@ -62,36 +63,52 @@ class Meta:
...
-    comments = fields.List(fields.Nested(EditingRevisionCommentSchema))
+    comments = fields.Method('_get_comments')
     initial_state = fields.Nested(RevisionState)
     final_state = fields.Nested(RevisionState)
+    create_comment_url = fields.Function(lambda revision: url_for('event_editing.api_create_comment', revision))
+
+    def _get_comments(self, revision):
+        comments = []
+        current_user = self.context.get('user')
+        contribution = revision.editable.contribution
```
Comment (ThiefMaster) [inline indico/modules/events/editing/schemas.py]: i'd add a property to the Editable that returns the contribution's event

File: indico/modules/events/editing/schemas.py [hunk 13]
```diff
@@ -62,36 +63,52 @@ class Meta:
...
     final_state = fields.Nested(RevisionState)
+    create_comment_url = fields.Function(lambda revision: url_for('event_editing.api_create_comment', revision))
+
+    def _get_comments(self, revision):
+        comments = []
+        current_user = self.context.get('user')
+        event = revision.editable.event
+        comments = [comment for comment in revision.comments
+                    if not comment.internal
+                    or event.can_manage(current_user, permission='paper_editing')]
```
Comment (ThiefMaster) [inline indico/modules/events/editing/schemas.py]: this fits in the previous line now :)

**Pull request merged**